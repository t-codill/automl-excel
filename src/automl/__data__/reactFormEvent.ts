// tslint:disable: no-any
export const reactFormEvent: React.FormEvent<any> = {
    preventDefault(): void { return; }
} as unknown as React.FormEvent<any>;
