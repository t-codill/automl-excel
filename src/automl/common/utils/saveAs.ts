export const saveAs = (content: Blob, fileName: string): void => {
    const evt = document.createEvent("MouseEvents");
    evt.initMouseEvent("click", true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);

    const link = window.document.createElement("a");
    link.setAttribute("href", window.URL.createObjectURL(content));
    link.setAttribute("download", fileName);
    link.dispatchEvent(evt);
};
