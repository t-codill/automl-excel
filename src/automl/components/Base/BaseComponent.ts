import { mapValues } from "lodash";
import { BasicTypes } from "../../common/BasicTypes";
import { IDictionary } from "../../common/IDictionary";
import { IServiceBaseProps } from "../../services/ServiceBase";
import { BaseComponentContext } from "./BaseComponentContext";

export abstract class BaseComponent<
    TProp,
    TState,
    TServices extends {
        [key: string]: {
            dispose(): void;
            reset(): void;
        };
    }> extends BaseComponentContext<TProp, TState> {

    protected get autoRefreshEnabled(): boolean | undefined {
        return this._autoRefreshEnabled;
    }

    protected set autoRefreshEnabled(flag: boolean | undefined) {
        this._autoRefreshEnabled = flag;
    }

    protected services!: { [key in keyof TServices]: TServices[key] };

    protected abstract readonly serviceConstructors: { [key in keyof TServices]: (new (context: IServiceBaseProps, props: TProp) => TServices[key]) };

    protected abstract readonly getData?: () => Promise<void>;

    protected autoRefreshInMs: number | undefined;

    protected refreshTimeOut: number | undefined;

    private mounted = false;
    private _autoRefreshEnabled: boolean | undefined;

    constructor(props: TProp) {
        super(props);
        this.setState = this.setState.bind(this);
        this.refresh = this.refresh.bind(this);
        this.changeAutoRefresh = this.changeAutoRefresh.bind(this);
    }

    public async componentDidMount(): Promise<void> {
        this.mounted = true;
        this.services = mapValues(this.serviceConstructors, (cons) => {
            return new cons(this.context, this.props);
        }) as { [key in keyof TServices]: TServices[key] };
        await this.refresh();
    }

    public componentWillUnmount(): void {
        this.mounted = false;
        if (this.refreshTimeOut) {
            clearTimeout(this.refreshTimeOut);
        }
        Object.keys(this.services)
            .forEach((key) => {
                this.services[key].dispose();
            });
    }

    public setState<K extends keyof TState>(
        state: ((prevState: Readonly<TState>, props: Readonly<TProp>) => (Pick<TState, K> | TState | null)) | (Pick<TState, K> | TState | null),
        callback?: () => void
    ): void {
        if (this.mounted) {
            super.setState(state, callback);
        }
    }

    protected logUserAction(component: string, properties?: IDictionary<BasicTypes>): void {
        this.context.logger.logCustomEvent(
            `${this.context.pageName}_${component}_UserAction`,
            this.context,
            {
                ...properties,
                pageName: this.context.pageName,
                component
            }
        );
    }

    protected async refresh(): Promise<void> {
        if (!this.mounted) {
            return;
        }
        if (this.getData) {
            if (this.refreshTimeOut) {
                clearTimeout(this.refreshTimeOut);
            }
            this.resetService();
            this.context.setLoading(true);
            await this.getData();
            this.context.setLoading(false);
            if (this.mounted && this._autoRefreshEnabled && this.autoRefreshInMs) {
                this.refreshTimeOut = window.setTimeout(this.refresh.bind(this), this.autoRefreshInMs);
            }
        }
    }

    protected changeAutoRefresh(enable: boolean): void {
        if (!this.autoRefreshInMs) {
            return;
        }
        if (this.refreshTimeOut) {
            clearTimeout(this.refreshTimeOut);
        }
        if (!this.mounted) {
            return;
        }
        if (enable) {
            this.refreshTimeOut = window.setTimeout(this.refresh.bind(this), this.autoRefreshInMs);
        }
        this._autoRefreshEnabled = enable;
    }

    private resetService(): void {
        Object.keys(this.services)
            .forEach((key) => {
                this.services[key].reset();
            });
    }
}
