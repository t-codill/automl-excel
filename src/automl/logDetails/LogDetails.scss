.log-details-container {
    display: flex;
    flex-direction: column;
    height: 100%;
    
    .log-content-container {
        flex: 1 1 auto;
    }
}