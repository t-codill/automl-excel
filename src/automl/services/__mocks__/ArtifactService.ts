import { ArtifactAPIModels } from "@vienna/artifact";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export class ArtifactService {
    private readonly tryGetContentResult = "tryGetContent";
    private readonly getAllArtifactsResult: ArtifactAPIModels.ArtifactDto[] = [
        {
            artifactId: "artifactId1",
            container: "container1",
            origin: "origin1",
            path: "path1"
        },
        {
            artifactId: "artifactId2",
            container: "container2",
            origin: "origin2",
            path: "path2"
        }
    ];
    private readonly tryGetArtifactUrlResult: ArtifactAPIModels.ArtifactContentInformationDto = {
        contentUri: "contentUri",
        path: "path"
    };
    private readonly getModelUrlResult = "getModelUrl";
    private readonly getModelUrlsResult = ["getModelUrls1", "getModelUrls2"];

    public async tryGetContentForRun(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async tryGetContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return params.map((p) => `"getAllContentsResult:${p.origin}/${p.container}/${p.path}`);
    }

    public async getAllArtifacts(): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllArtifactsResult;
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return runs.map(() => this.getAllArtifactsResult);
    }

    public async tryGetArtifactUrl(): Promise<ArtifactAPIModels.ArtifactContentInformationDto | null | undefined> {
        return this.tryGetArtifactUrlResult;
    }

    public async getModelUrl(): Promise<string | undefined> {
        return this.getModelUrlResult;
    }

    public async getModelUrls(): Promise<Array<string | null> | undefined> {
        return this.getModelUrlsResult;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
