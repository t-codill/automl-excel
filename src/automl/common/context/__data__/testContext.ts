import { PageNames } from "../../PageNames";
import { Logger } from "../../utils/logger";
import { IWorkspaceContextProps, WorkspaceError } from "../IWorkspaceProps";
import { WorkspaceFlight } from "../WorkspaceFlight";

export function getLogCustomEventSpy(): jest.SpyInstance<ReturnType<Logger["logCustomEvent"]>> {
    const spy = jest.spyOn(Logger.prototype, "logCustomEvent");
    spy.mockImplementation(jest.fn());
    return spy;
}
let pageName: PageNames = PageNames.Unknown;
export const testContext: IWorkspaceContextProps = {
    get pageName(): PageNames { return pageName; },
    get theme(): string { return "light"; },
    discoverUrls: {},
    flight: new WorkspaceFlight(""),
    location: "eastus",
    resourceGroupName: "testResource",
    subscriptionId: "00000000-0000-0000-0000-000000000000",
    logger: new Logger("test"),
    getToken: () => "testToken",
    setPageName: (value: PageNames) => { pageName = value; },
    workspaceName: "testWorkSpace",
    setNavigationBarButtons(): void { return; },
    onError: (err: WorkspaceError) => { throw err; },
    clearErrors: () => { return; },
    setHeader: () => { return; },
    setLoading: () => { return; }
};
