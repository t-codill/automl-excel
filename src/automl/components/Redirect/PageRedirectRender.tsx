import * as React from "react";
import { generatePath, Redirect } from "react-router-dom";
import { Main } from "../../Main";
import { BaseComponent } from "../Base/BaseComponent";

export class PageRedirectRender extends BaseComponent<{
    noPush: boolean;
    expendedRoutePath: string;
}, {}, {}> {
    protected serviceConstructors = {};
    protected getData = undefined;
    public render(): React.ReactNode {
        if (this.props.noPush) {
            return <Redirect to={this.context.flight.appendQueryParam(`${generatePath(Main.routePath, {
                resourceGroupName: this.context.resourceGroupName,
                subscriptionId: this.context.subscriptionId,
                workspaceName: this.context.workspaceName
            })}/${this.props.expendedRoutePath}`)} push={false} />;
        }
        else {
            return <Redirect to={this.context.flight.appendQueryParam(`${generatePath(Main.routePath, {
                resourceGroupName: this.context.resourceGroupName,
                subscriptionId: this.context.subscriptionId,
                workspaceName: this.context.workspaceName
            })}/${this.props.expendedRoutePath}`)} push={true} />;
        }

    }
}
