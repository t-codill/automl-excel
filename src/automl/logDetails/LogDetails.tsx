import { ArtifactAPIModels } from "@vienna/artifact";
import JSZip from "jszip";
import { Dropdown, DropdownMenuItemType, ICommandBarItemProps, IDropdownOption, MessageBar } from "office-ui-fabric-react";
import path from "path";
import * as React from "react";
import { ChildRun } from "../childRun/ChildRun";
import { compare } from "../common/compare/index";
import { IDataLoadState } from "../common/IDataLoadState";
import { PageNames } from "../common/PageNames";
import { nameof } from "../common/utils/nameof";
import { saveAs } from "../common/utils/saveAs";
import { BasePage } from "../components/Base/BasePage";
import { PageLoadingSpinner } from "../components/Progress/PageLoadingSpinner";
import { PageRedirect } from "../components/Redirect/PageRedirect";
import { ParentRun } from "../parentRun/ParentRun";
import { ArtifactService, IArtifact } from "../services/ArtifactService";
import { RunLogContent } from "./RunLogContent";

import "./LogDetails.scss";

export interface ILogDetailsRouteProps {
    experimentName: string;
    runId: string;
    runType: "parent" | "child";
}

export interface IRunLog extends IArtifact {
    logName: string;
    prefix: string;
}

export interface ILogDetailsState extends IDataLoadState {
    selectedOption: IDropdownOption | undefined;
    logs: IRunLog[];
    logOptions: IDropdownOption[];
    refreshTimestamp: number;
}

export class LogDetails extends BasePage<
    ILogDetailsRouteProps,
    ILogDetailsState,
    { artifactService: ArtifactService }> {
    public static routePath = `logs/:${nameof<ILogDetailsRouteProps>("experimentName")}/LogDetails/:${nameof<ILogDetailsRouteProps>("runType")}/:${nameof<ILogDetailsRouteProps>("runId")}`;
    protected readonly header = "Log Detail";
    protected readonly pageName = PageNames.LogDetails;
    protected readonly serviceConstructors = { artifactService: ArtifactService };
    protected readonly navigationBarButtons: ICommandBarItemProps[] = [
        {
            key: "download",
            text: "Download All",
            iconProps: {
                iconName: "Download"
            },
            onClick: () => {
                this.downloadAll();
            }
        }
    ];
    private readonly logsRegex = /(^azureml-logs\/|\.log$)/;

    constructor(props: ILogDetailsRouteProps) {
        super(props);

        this.state = {
            goBack: false,
            isDataLoaded: false,
            selectedOption: undefined,
            logs: [],
            logOptions: [],
            refreshTimestamp: Date.now()
        };
    }
    public readonly render = (): React.ReactNode => {
        if (this.state.goBack) {
            if (this.props.runType === "parent") {
                return PageRedirect(ParentRun, { experimentName: this.props.experimentName, runId: this.props.runId });
            }
            else {
                return PageRedirect(ChildRun, { experimentName: this.props.experimentName, runId: this.props.runId });
            }
        }
        if (!this.state.isDataLoaded) {
            return <PageLoadingSpinner />;
        }
        if (this.state.logOptions.length < 1) {
            return this.renderNoLogs();
        }

        return <div className="log-details-container">
            <div className="log-dropdown-container">
                <Dropdown
                    selectedKey={this.state.selectedOption && this.state.selectedOption.key}
                    onChange={this.onDropDownChange}
                    options={this.state.logOptions}
                />
            </div>
            <div className="log-content-container">
                {
                    this.state.selectedOption &&
                    <RunLogContent
                        key={this.state.selectedOption.data.path}
                        refreshTimestamp={this.state.refreshTimestamp}
                        runLog={this.state.selectedOption.data} />
                }
            </div>

        </div>;
    }

    protected readonly getData = async () => {
        const setupRunId = `${this.props.runId}_setup`;
        const allArtifacts = await this.services.artifactService.getAllArtifactsForRuns([this.props.runId, setupRunId]);

        if (!allArtifacts) {
            return;
        }

        const [artifacts, setupArtifacts] = allArtifacts;

        const runLogs = this.getLogs(artifacts);
        const runLogOptions = this.generateLogOptions(runLogs);

        const setupLogs = this.getLogs(setupArtifacts, "setup_");
        const setupLogOptions = this.generateLogOptions(setupLogs);

        const logs = [...runLogs, ...setupLogs];
        const logOptions = [...runLogOptions, ...setupLogOptions];

        const selectedOption = this.state.selectedOption || logOptions.find((logOption) => logOption.itemType !== DropdownMenuItemType.Header);
        const downloadButton = this.navigationBarButtons.find((button) => {
            return button.key === "download";
        });
        if (downloadButton) {
            downloadButton.disabled = logs.length < 1;
        }

        this.setState({
            isDataLoaded: true,
            selectedOption,
            logs,
            logOptions,
            refreshTimestamp: Date.now()
        });
    }

    private readonly renderNoLogs = () => {
        return <MessageBar>
            No logs are found.
        </MessageBar>;
    }

    private readonly getLogs = (artifacts: ArtifactAPIModels.ArtifactDto[], prefix = "") => {
        const logs = artifacts
            .map((artifact) => ({
                logName: prefix + path.basename(artifact.path || ""),
                prefix,
                path: artifact.path || "",
                container: artifact.container || "",
                origin: artifact.origin || ""
            }));
        return logs.sort((a: IRunLog, b: IRunLog) => compare(a.path, b.path));
    }

    private readonly generateLogOptions = (logs: IRunLog[]) => {
        const logOptions: IDropdownOption[] = [];
        let lastHeader = "";

        for (const log of logs
            .filter((artifact) => artifact.path && this.logsRegex.test(artifact.path))) {
            const header = this.getFolderPath(log.path);
            if (header !== undefined && header !== "" && header !== lastHeader) {
                lastHeader = header;
                logOptions.push(
                    {
                        key: `header-${log.path}`,
                        text: header,
                        itemType: DropdownMenuItemType.Header
                    }
                );
            }
            logOptions.push(
                {
                    key: log.path,
                    text: `-> ${log.logName}`,
                    title: log.logName,
                    data: log
                }
            );
        }
        return logOptions;
    }

    private readonly getFolderPath = (fullPath: string): string | undefined => {
        const paths: string[] = fullPath.split("/");
        paths.splice(-1, 1);
        return paths.join("/");
    }

    private readonly onDropDownChange = (_event: React.FormEvent<HTMLDivElement>, item?: IDropdownOption) => {
        if (!item) {
            return;
        }
        this.setState({ selectedOption: item });
    }

    private readonly downloadAll = async () => {
        const contents = await this.services.artifactService.getAllContents(this.state.logs);
        if (!contents) {
            return;
        }
        this.logUserAction("Download All", { runId: this.props.runId });
        const zip = new JSZip();
        for (let i = 0; i < this.state.logs.length; i++) {
            const content = contents[i];
            if (content) {
                zip.file(this.state.logs[i].prefix ? path.join(this.state.logs[i].prefix, this.state.logs[i].path) : this.state.logs[i].path, content);
            }
        }
        const zipContent = await zip.generateAsync({ type: "blob" });
        saveAs(zipContent, "log.zip");
    }
}
