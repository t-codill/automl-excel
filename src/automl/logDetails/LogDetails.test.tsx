import { ArtifactAPIModels } from "@vienna/artifact";
import { shallow } from "enzyme";
import JSZip from "jszip";
import { Dropdown, ICommandBarItemProps } from "office-ui-fabric-react";
import * as React from "react";
import { reactFormEvent } from "../__data__/reactFormEvent";
import { getLogCustomEventSpy, testContext } from "../common/context/__data__/testContext";
import * as saveAsModule from "../common/utils/saveAs";
import { BaseComponent } from "../components/Base/BaseComponent";
import { PageRedirectRender } from "../components/Redirect/PageRedirectRender";
import { ArtifactService } from "../services/ArtifactService";
import { LogDetails } from "./LogDetails";

jest.mock("../services/ArtifactService");

let asSaveAs: jest.SpyInstance<void, [Blob, string]>;
let asGetAllContents: jest.SpyInstance<ReturnType<ArtifactService["getAllContents"]>>;
let asGetAllArtifactsForRuns: jest.SpyInstance<ReturnType<ArtifactService["getAllArtifactsForRuns"]>>;
let asJSZipFile: jest.SpyInstance<ReturnType<JSZip["file"]>>;

const renderLogDetailsWithMockArtifacts = async (artifacts: ArtifactAPIModels.ArtifactDto[][] | undefined) => {
    const mockArtifacts = Promise.resolve(artifacts);
    asGetAllArtifactsForRuns.mockReturnValue(mockArtifacts);
    let buttons: ICommandBarItemProps[] = [];
    jest.spyOn(BaseComponent.prototype.context, "setNavigationBarButtons")
        .mockImplementation((b) => { buttons = b; });
    const wrapper = shallow<LogDetails>(<LogDetails runId="run_id" experimentName="experiment_name" runType="parent" />);
    await mockArtifacts;
    const downloadAll = buttons.find((b) => b.key === "download");
    return { wrapper, downloadAll };
};

describe("LogDetails", () => {
    const logs = [
        {
            artifactId: "empty.log",
            path: undefined
        },
        {
            artifactId: "log.log",
            path: "log.log"
        },
        {
            artifactId: "60_control_log.txt",
            path: "azureml-logs/60_control_log.txt"
        },
        {
            artifactId: "80_driver_log.txt",
            path: "azureml-logs/80_driver_log.txt"
        }
    ];

    beforeEach(() => {
        asJSZipFile = jest.spyOn(JSZip.prototype, "file");
        asSaveAs = jest.spyOn(saveAsModule, "saveAs");
        asGetAllContents = jest.spyOn(ArtifactService.prototype, "getAllContents");
        asGetAllArtifactsForRuns = jest.spyOn(ArtifactService.prototype, "getAllArtifactsForRuns");
    });

    describe("go back", () => {
        it("to parent", () => {
            const wrapper = shallow(<LogDetails runId="run_id" experimentName="experiment_name" runType="parent" />);
            wrapper.setState({ goBack: true });
            expect(wrapper.find(PageRedirectRender)
                .props())
                .toEqual({
                    expendedRoutePath: "experiments/experiment_name/parentrun/run_id",
                    noPush: false,
                });
        });

        it("to child", () => {
            const wrapper = shallow(<LogDetails runId="run_id" experimentName="experiment_name" runType="child" />);
            wrapper.setState({ goBack: true });
            expect(wrapper.find(PageRedirectRender)
                .props())
                .toEqual({
                    expendedRoutePath: "experiments/experiment_name/childrun/run_id",
                    noPush: false,
                });
        });
    });

    describe("render", () => {

        it("should render page loading spinner by default", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts(undefined);
            expect(wrapper)
                .toMatchInlineSnapshot("<pageLoadingSpinner />");
        });

        it("should render empty message if get all artifacts returns empty", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts([[], []]);
            expect(wrapper)
                .toMatchSnapshot();
        });

        it("should render with only run logs", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts([logs, []]);
            expect(wrapper)
                .toMatchSnapshot();
        });

        it("should render with only setup run logs", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts([[], logs]);
            expect(wrapper)
                .toMatchSnapshot();
        });

        it("should render with both run and setup run logs", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts([logs, logs]);
            expect(wrapper)
                .toMatchSnapshot();
        });
    });

    describe("change file", () => {
        it("should ignore if no selected", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts([logs, []]);
            const ddWrapper = wrapper.find(Dropdown);
            const ddProps = ddWrapper.props();
            if (ddProps && ddProps.onChange) {
                ddProps.onChange(reactFormEvent, undefined);
            }
            expect(wrapper.state("selectedOption"))
                .toBeDefined();
        });

        it("should change selected option", async () => {
            const { wrapper } = await renderLogDetailsWithMockArtifacts([logs, []]);
            const onChange = wrapper
                .find(Dropdown)
                .props()
                .onChange;

            const option = {
                key: "azureml-logs/80_driver_log.txt",
                text: "-> 80_driver_log.txt",
                data: logs[0]
            };
            if (onChange) {
                onChange(reactFormEvent, option);
            }
            expect(wrapper.state("selectedOption"))
                .toEqual(option);
        });
    });

    describe("download", () => {
        it("should disable download button", async () => {
            const { downloadAll } = await renderLogDetailsWithMockArtifacts([[], []]);
            let disabled: boolean | undefined;
            if (downloadAll) {
                disabled = downloadAll.disabled;
            }
            expect(disabled)
                .toBe(true);
        });
        it("should not disable download button if not found", async () => {
            const mockArtifacts = Promise.resolve([logs, []]);
            asGetAllArtifactsForRuns.mockReturnValue(mockArtifacts);
            let downloadAll: ICommandBarItemProps | undefined;
            jest.spyOn(BaseComponent.prototype.context, "setNavigationBarButtons")
                .mockImplementation((buttons) => {
                    downloadAll = buttons.find((b) => b.key === "download");
                    if (downloadAll) {
                        downloadAll.key = "downloadBak";
                    }
                });
            shallow<LogDetails>(<LogDetails runId="run_id" experimentName="experiment_name" runType="parent" />);
            await mockArtifacts;
            let disabled: boolean | undefined;
            if (downloadAll) {
                disabled = downloadAll.disabled;
            }
            expect(disabled)
                .toBeUndefined();
        });
        it("should download all", async () => {
            const { downloadAll } = await renderLogDetailsWithMockArtifacts([logs, logs]);
            const logSpy = getLogCustomEventSpy();
            if (downloadAll && downloadAll.onClick) {
                downloadAll.onClick();
            }
            await Promise.resolve();
            await Promise.resolve();
            expect(logSpy)
                .toBeCalledWith(
                    "LogDetails_Download All_UserAction",
                    testContext,
                    { component: "Download All", pageName: "LogDetails", runId: "run_id" }
                );
            expect(asSaveAs)
                .toBeCalledWith(expect.anything(), "log.zip");
        });
        it("should not download if return content is undefined", async () => {
            asGetAllContents.mockReturnValue(Promise.resolve(undefined));
            const { downloadAll } = await renderLogDetailsWithMockArtifacts([logs, logs]);
            if (downloadAll && downloadAll.onClick) {
                downloadAll.onClick();
            }
            await Promise.resolve();
            await Promise.resolve();
            expect(asSaveAs)
                .toHaveBeenCalledTimes(0);
        });
        it("should not add empty files", async () => {
            asGetAllContents.mockReturnValue(Promise.resolve([undefined, undefined]));
            const { downloadAll } = await renderLogDetailsWithMockArtifacts([logs, logs]);
            const logSpy = getLogCustomEventSpy();
            if (downloadAll && downloadAll.onClick) {
                downloadAll.onClick();
            }
            await Promise.resolve();
            await Promise.resolve();
            expect(logSpy)
                .toBeCalledWith(
                    "LogDetails_Download All_UserAction",
                    testContext,
                    { component: "Download All", pageName: "LogDetails", runId: "run_id" }
                );
            expect(asJSZipFile)
                .toHaveBeenCalledTimes(0);
            expect(asSaveAs)
                .toBeCalledWith(expect.anything(), "log.zip");
        });
    });
});
