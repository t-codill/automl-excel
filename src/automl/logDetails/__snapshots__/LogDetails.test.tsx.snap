// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogDetails render should render empty message if get all artifacts returns empty 1`] = `
<StyledMessageBarBase>
  No logs are found.
</StyledMessageBarBase>
`;

exports[`LogDetails render should render with both run and setup run logs 1`] = `
<div
  className="log-details-container"
>
  <div
    className="log-dropdown-container"
  >
    <StyledWithResponsiveMode
      onChange={[Function]}
      options={
        Array [
          Object {
            "itemType": 2,
            "key": "header-azureml-logs/60_control_log.txt",
            "text": "azureml-logs",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "60_control_log.txt",
              "origin": "",
              "path": "azureml-logs/60_control_log.txt",
              "prefix": "",
            },
            "key": "azureml-logs/60_control_log.txt",
            "text": "-> 60_control_log.txt",
            "title": "60_control_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "80_driver_log.txt",
              "origin": "",
              "path": "azureml-logs/80_driver_log.txt",
              "prefix": "",
            },
            "key": "azureml-logs/80_driver_log.txt",
            "text": "-> 80_driver_log.txt",
            "title": "80_driver_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "log.log",
              "origin": "",
              "path": "log.log",
              "prefix": "",
            },
            "key": "log.log",
            "text": "-> log.log",
            "title": "log.log",
          },
          Object {
            "itemType": 2,
            "key": "header-azureml-logs/60_control_log.txt",
            "text": "azureml-logs",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "setup_60_control_log.txt",
              "origin": "",
              "path": "azureml-logs/60_control_log.txt",
              "prefix": "setup_",
            },
            "key": "azureml-logs/60_control_log.txt",
            "text": "-> setup_60_control_log.txt",
            "title": "setup_60_control_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "setup_80_driver_log.txt",
              "origin": "",
              "path": "azureml-logs/80_driver_log.txt",
              "prefix": "setup_",
            },
            "key": "azureml-logs/80_driver_log.txt",
            "text": "-> setup_80_driver_log.txt",
            "title": "setup_80_driver_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "setup_log.log",
              "origin": "",
              "path": "log.log",
              "prefix": "setup_",
            },
            "key": "log.log",
            "text": "-> setup_log.log",
            "title": "setup_log.log",
          },
        ]
      }
      selectedKey="azureml-logs/60_control_log.txt"
    />
  </div>
  <div
    className="log-content-container"
  >
    <RunLogContent
      key="azureml-logs/60_control_log.txt"
      refreshTimestamp={0}
      runLog={
        Object {
          "container": "",
          "logName": "60_control_log.txt",
          "origin": "",
          "path": "azureml-logs/60_control_log.txt",
          "prefix": "",
        }
      }
    />
  </div>
</div>
`;

exports[`LogDetails render should render with only run logs 1`] = `
<div
  className="log-details-container"
>
  <div
    className="log-dropdown-container"
  >
    <StyledWithResponsiveMode
      onChange={[Function]}
      options={
        Array [
          Object {
            "itemType": 2,
            "key": "header-azureml-logs/60_control_log.txt",
            "text": "azureml-logs",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "60_control_log.txt",
              "origin": "",
              "path": "azureml-logs/60_control_log.txt",
              "prefix": "",
            },
            "key": "azureml-logs/60_control_log.txt",
            "text": "-> 60_control_log.txt",
            "title": "60_control_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "80_driver_log.txt",
              "origin": "",
              "path": "azureml-logs/80_driver_log.txt",
              "prefix": "",
            },
            "key": "azureml-logs/80_driver_log.txt",
            "text": "-> 80_driver_log.txt",
            "title": "80_driver_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "log.log",
              "origin": "",
              "path": "log.log",
              "prefix": "",
            },
            "key": "log.log",
            "text": "-> log.log",
            "title": "log.log",
          },
        ]
      }
      selectedKey="azureml-logs/60_control_log.txt"
    />
  </div>
  <div
    className="log-content-container"
  >
    <RunLogContent
      key="azureml-logs/60_control_log.txt"
      refreshTimestamp={0}
      runLog={
        Object {
          "container": "",
          "logName": "60_control_log.txt",
          "origin": "",
          "path": "azureml-logs/60_control_log.txt",
          "prefix": "",
        }
      }
    />
  </div>
</div>
`;

exports[`LogDetails render should render with only setup run logs 1`] = `
<div
  className="log-details-container"
>
  <div
    className="log-dropdown-container"
  >
    <StyledWithResponsiveMode
      onChange={[Function]}
      options={
        Array [
          Object {
            "itemType": 2,
            "key": "header-azureml-logs/60_control_log.txt",
            "text": "azureml-logs",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "setup_60_control_log.txt",
              "origin": "",
              "path": "azureml-logs/60_control_log.txt",
              "prefix": "setup_",
            },
            "key": "azureml-logs/60_control_log.txt",
            "text": "-> setup_60_control_log.txt",
            "title": "setup_60_control_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "setup_80_driver_log.txt",
              "origin": "",
              "path": "azureml-logs/80_driver_log.txt",
              "prefix": "setup_",
            },
            "key": "azureml-logs/80_driver_log.txt",
            "text": "-> setup_80_driver_log.txt",
            "title": "setup_80_driver_log.txt",
          },
          Object {
            "data": Object {
              "container": "",
              "logName": "setup_log.log",
              "origin": "",
              "path": "log.log",
              "prefix": "setup_",
            },
            "key": "log.log",
            "text": "-> setup_log.log",
            "title": "setup_log.log",
          },
        ]
      }
      selectedKey="azureml-logs/60_control_log.txt"
    />
  </div>
  <div
    className="log-content-container"
  >
    <RunLogContent
      key="azureml-logs/60_control_log.txt"
      refreshTimestamp={0}
      runLog={
        Object {
          "container": "",
          "logName": "setup_60_control_log.txt",
          "origin": "",
          "path": "azureml-logs/60_control_log.txt",
          "prefix": "setup_",
        }
      }
    />
  </div>
</div>
`;
